File: solutions/[139]单词拆分.cpp
#include <string>
#include <vector>
#include <set>
using namespace std;

class Solution {
public:
	//动态规划，自底向上
	bool wordBreak(string s, vector<string>& wordDict) {
		vector<bool> memo(s.length() + 1, false);
		memo[0] = true;
		set<string> ws(wordDict.begin(), wordDict.end());

		for (int i = 1; i <= s.length(); ++i) {
			for (int j = 0; j < i; ++j) {
				if (memo[j] && ws.find(s.substr(j, i - j)) != ws.end()) {
					memo[i] = true;
					break;
				}
			}
		}
		return memo[s.length()];
	}

	//递归写法
};

File: solutions/[139]单词拆分.py
from typing import List


class Trie:
    def __init__(self, word_list=None) -> None:
        self.count = 0
        self.next = [[0] * 26 for _ in range(10000)]
        self.is_end = [False] * 10000
        if word_list is not None:
            for word in word_list:
                self.insert(word)

    def insert(self, word):
        p = 0
        for c in word:
            c = ord(c) - ord('a')
            if self.next[p][c] == 0:
                self.count += 1
                self.next[p][c] = self.count
            p = self.next[p][c]
        self.is_end[p] = True
        # print(word, self.is_end[p], p)

    def find(self, word):
        p = 0
        for c in word:
            # print(word, c)
            c = ord(c) - ord('a')
            if self.next[p][c] == 0:
                return False
            p = self.next[p][c]
        # print('find:', word, self.is_end[p], p)
        return self.is_end[p]

class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        trie = Trie(wordDict)
        trie.insert('')
        memo = [False] * (len(s)+1)
        memo[0] = True

        for i in range(1, len(s) + 1):
            for j in range(i):
                memo[i] = memo[i] or (memo[j] and trie.find(s[j:i])
            )
        # print(memo)
        # print(trie.find(''))
        # for word in wordDict:
        #     print(trie.find(word))
        return memo[len(s)]
